lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+        message (str): message to be output
+    """
+    cwrite('@*g{[+]} ')
+    print(message)
+
+
+def _process_external_package(pkg, explicit):
+    """
+    Helper function to run post install hooks and register external packages.
+
+    Args:
+        pkg (Package): the external package
+        explicit (bool): if the package was requested explicitly by the user,
+            ``False`` if it was pulled in as a dependency of an explicit
+            package.
+    """
+    # TODO: Should this error out?
+    if not pkg.spec.external:
+        return
+
+    pre = '{s.name}@{s.version} :'.format(s=pkg.spec)
+    spec = pkg.spec
+
+    if spec.external_module:
+        tty.msg('{0} has external module in {1}'
+                .format(pre, spec.external_module))
+        tty.msg('{0} is actually installed in {1}'
+                .format(pre, spec.external_path))
+    else:
+        tty.msg("{0} externally installed in {1}"
+                .format(pre, spec.external_path))
+
+    try:
+        # Check if the package was already registered in the DB.
+        # If this is the case, then just exit.
+        rec = spack.store.db.get_record(spec)
+        tty.msg('{0} already registered in DB'.format(pre))
+
+        # Update the value of rec.explicit if it is necessary
+        _update_explicit_entry_in_db(pkg, rec, explicit)
+
+    except KeyError:
+        # If not, register it and generate the module file.
+        # For external packages we just need to run
+        # post-install hooks to generate module files.
+        tty.msg('{0} generating module file'.format(pre))
+        spack.hooks.post_install(spec)
+
+        # Add to the DB
+        tty.msg('{0} registering into DB'.format(pre))
+        spack.store.db.add(spec, None, explicit=explicit)
+
+
+def _try_install_from_binary_cache(pkg, explicit):
+    """
+    Try to install the package from binary cache.
+
+    Args:
+        pkg (spack.package.PackageBase): the package to be installed from
+            binary cache
+        explicit (bool): the package was explicitly requested by the user
+    """
+    tty.debug('Searching for binary cache of {0}'.format(pkg.unique_id))
+    specs = binary_distribution.get_specs()
+    binary_spec = spack.spec.Spec.from_dict(pkg.spec.to_dict())
+    binary_spec._mark_concrete()
+    if binary_spec not in specs:
+        return False
+
+    tarball = binary_distribution.download_tarball(binary_spec)
+    # see #10063 : install from source if tarball doesn't exist
+    if tarball is None:
+        tty.msg('{0} exists in binary cache but with different hash'
+                .format(pkg.name))
+        return False
+
+    tty.msg('Installing {0} from binary cache'.format(pkg.unique_id))
+    binary_distribution.extract_tarball(
+        binary_spec, tarball, allow_root=False,
+        unsigned=False, force=False)
+    pkg.installed_from_binary_cache = True
+    spack.store.db.add(pkg.spec, spack.store.layout, explicit=explicit)
+    return True
```

Comment:
For instance this seems exactly the same as:

https://github.com/spack/spack/blob/ca26eb69231fffb500459746a218b04724657652/lib/spack/spack/package.py#L1508-L1528

Reply:
Yes, pretty much.  

This one is no longer in `package` in part because the `do_install` process is no longer recursive; rather, it needs to potentially be called for each package as it is popped from priority queue.